### ISSUE TITLE
Data::Dumper: Dumper function fails to 'round-trip' when strict vars is on

### ISSUE THREAD

**Issue by jkeenan** [atoomic/perl#192]
Consider the following Perl 5 program, which is adapted from the core distribution test suite:
```
$ cat dd.t
use Test::More;
use Data::Dumper;

{
    my $q = q| \/ |;
    my $qr = qr{$q};
    eval Dumper $qr;
    ok(!$@, "Dumping $qr with XS") or diag $@, Dumper $qr;
    local $Data::Dumper::Useperl = 1;
    eval Dumper $qr;
    ok(!$@, "Dumping $qr with PP") or diag $@, Dumper $qr;
}

done_testing();
```
When we run it as is, we get:
```
$ prove -v dd.t
dd.t .. 
ok 1 - Dumping (?^: \/ ) with XS
ok 2 - Dumping (?^: \/ ) with PP
1..2
ok
All tests successful.
```
But suppose we ask, *"What happens if we run this under `use strict`?"*  We revise the file to be this:
```
$ cat dd.t
use Test::More;
use Data::Dumper;

{
    no strict;
    my $q = q| \/ |;
    my $qr = qr{$q};
    eval Dumper $qr;
    ok(!$@, "Dumping $qr with XS") or diag $@, Dumper $qr;
    local $Data::Dumper::Useperl = 1;
    eval Dumper $qr;
    ok(!$@, "Dumping $qr with PP") or diag $@, Dumper $qr;
}

{
    use strict 'vars';
    my $q = q| \/ |;
    my $qr = qr{$q};
    eval Dumper $qr;
    ok(!$@, "Dumping $qr with XS") or diag $@, Dumper $qr;
    local $Data::Dumper::Useperl = 1;
    eval Dumper $qr;
    ok(!$@, "Dumping $qr with PP") or diag $@, Dumper $qr;
}

done_testing();
```
Let's run it.
```
$ prove -v dd.t
dd.t .. 
ok 1 - Dumping (?^: \/ ) with XS
ok 2 - Dumping (?^: \/ ) with PP
not ok 3 - Dumping (?^: \/ ) with XS
Variable "$VAR1" is not imported at (eval 10) line 1.

#   Failed test 'Dumping (?^: \/ ) with XS'
#   at dd.t line 20.
# Global symbol "$VAR1" requires explicit package name (did you forget to declare "my $VAR1"?) at (eval 10) line 1.
# $VAR1 = qr/ \/ /;
Variable "$VAR1" is not imported at (eval 11) line 1.
not ok 4 - Dumping (?^: \/ ) with PP

#   Failed test 'Dumping (?^: \/ ) with PP'
#   at dd.t line 23.
# Global symbol "$VAR1" requires explicit package name (did you forget to declare "my $VAR1"?) at (eval 11) line 1.
# $VAR1 = qr/ \/ /;
1..4
# Looks like you failed 2 tests of 4.
Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/4 subtests 

Test Summary Report
-------------------
dd.t (Wstat: 512 Tests: 4 Failed: 2)
  Failed tests:  3-4
  Non-zero exit status: 2
Files=1, Tests=4,  1 wallclock secs ( 0.02 usr  0.00 sys +  0.05 cusr  0.00 csys =  0.07 CPU)
Result: FAIL
```
Uh-oh, when we impose strictures (actually, just one of the three kinds of strictures), the program fails.  That's because Data::Dumper's `Dumper` function (or the equivalent method call, `$d->Dump`) writes a string which describes a Perl data structure in which assignment is made to one or more **global** variables.  That's not a problem when we're just dumping to STDOUT for diagnostic purposes.  It is, however, a problem when we want to "round-trip" the output as a string to be `eval`-ed.  Under `use strict 'vars'`, the program being `eval`-ed fails because `$VAR1` is neither fully qualified nor scoped with `my`.

This round-tripping is described by Data::Dumper's documentation.
```
Data::Dumper - stringified perl data structures, suitable for both printing and C<eval>
...
The content of each variable is output in a single Perl statement. ... 
The return value can be C<eval>ed to get back an identical copy of the original reference structure.
```
In the original example above, the Perl statement produced would be:
```
$VAR1 = qr/ \/ /;
```
Now, while this is a problem in Perl 5, it's a much greater one in any version of Perl where strictures are on by default.  On the original version of the file, Perl 7 does this:
```
$ p7prove -v dd.t
dd.t .. 
not ok 1 - Dumping (?^: \/ ) with XS

#   Failed test 'Dumping (?^: \/ ) with XS'
#   at dd.t line 9.
# Global symbol "$VAR1" requires explicit package name (did you forget to declare "my $VAR1"?) at (eval 8) line 1.
# $VAR1 = qr/ \/ /;
not ok 2 - Dumping (?^: \/ ) with PP

#   Failed test 'Dumping (?^: \/ ) with PP'
#   at dd.t line 12.
# Global symbol "$VAR1" requires explicit package name (did you forget to declare "my $VAR1"?) at (eval 9) line 1.
# $VAR1 = qr/ \/ /;
1..2
# Looks like you failed 2 tests of 2.
Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/2 subtests 

Test Summary Report
-------------------
dd.t (Wstat: 512 Tests: 2 Failed: 2)
  Failed tests:  1-2
  Non-zero exit status: 2
```
In other words, it throws the same exceptions as Perl 5 does with strict-vars enabled.

How can we handle this?

Perhaps Data::Dumper could be revised to determine whether `use strict 'vars'` is in effect before it composes the string.  If strict-vars is in effect, it could prepend the variable declaration with `my`, thereby making it suitable for `eval`.  Something like this pseudo-code:
```
    my $qr = qr{$q};
    my $augmented = "my ". (Dumper $qr);
    eval $augmented;
```

Thank you very much.
Jim Keenan

**Comment by atoomic** [comment]
thanks for reporting, I will have a look at it later

**Comment by atoomic** [comment]
what about using a dummy helper like `add_my` in this example

```perl
#!perl

use Test::More;
use Data::Dumper;

{
    no strict;
    my $q = q| \/ |;
    my $qr = qr{$q};
    eval Dumper $qr;
    ok(!$@, "Dumping $qr with XS") or diag $@, Dumper $qr;
    local $Data::Dumper::Useperl = 1;
    eval Dumper $qr;
    ok(!$@, "Dumping $qr with PP") or diag $@, Dumper $qr;
}

{
    use strict 'vars';
    my $q = q| \/ |;
    my $qr = qr{$q};
    my $code = Dumper $qr;
    eval add_my( Dumper $qr );
    ok(!$@, "Dumping $qr with XS") or diag $@, Dumper $qr;
    local $Data::Dumper::Useperl = 1;
    eval add_my( Dumper $qr );
    ok(!$@, "Dumping $qr with PP") or diag $@, Dumper $qr;
}

done_testing();

sub add_my {
    my ( $str ) = @_;

    return unless defined $str;
    return $str =~ s{^(\s*\$VAR\d+\s+=)}{my $1}gam;
}
```

once run the result output looks like this
```
> perl5.30.0 dd.t
ok 1 - Dumping (?^: \/ ) with XS
ok 2 - Dumping (?^: \/ ) with PP
ok 3 - Dumping (?^: \/ ) with XS
ok 4 - Dumping (?^: \/ ) with PP
1..4
```

**Comment by jkeenan** [comment]
> what about using a dummy helper like `add_my` in this example
> 
> ```perl
> {
>     use strict 'vars';
>     my $q = q| \/ |;
>     my $qr = qr{$q};
>     my $code = Dumper $qr;
>     eval add_my( Dumper $qr );
>     ok(!$@, "Dumping $qr with XS") or diag $@, Dumper $qr;
>     local $Data::Dumper::Useperl = 1;
>     eval add_my( Dumper $qr );
>     ok(!$@, "Dumping $qr with PP") or diag $@, Dumper $qr;
> }
> 
> done_testing();
> 
> sub add_my {
>     my ( $str ) = @_;
> 
>     return unless defined $str;
>     return $str =~ s{^(\s*\$VAR\d+\s+=)}{my $1}gam;
> }
> ```
> 
> once run the result output looks like this
> 
> ```
> > perl5.30.0 dd.t
> ok 1 - Dumping (?^: \/ ) with XS
> ok 2 - Dumping (?^: \/ ) with PP
> ok 3 - Dumping (?^: \/ ) with XS
> ok 4 - Dumping (?^: \/ ) with PP
> 1..4
> ```

Granted, that would get the test to pass.  However, I'm raising a larger issue:  the principle of "least surprise."  Perl 5 programmers have presumably always expected to **not** have to pay attention to the scoping for `$VAR1`.  When you're using `Dumper` purely for diagnostic reasons during development, the scoping doesn't matter.  However, I suspect that once we're in strict-by-default country, where, essentially every Perl program in `STRING` in `eval STRING` must be strict as well, the scoping will be much more evident.

So what I'm suggesting is that we consider a change or an augmentation of Dumper's behavior.  If we do that, then we don't have to hack the tests.

Thank you very much.
Jim Keenan

**Comment by atoomic** [comment]
The problem of changing `Dumper` function behavior... is that it could break maybe unexpected code trying to parse the output. The alternate would be to provide an additional function (name to be defined) doing the job.

**Comment by jkeenan** [comment]
On 8/12/20 5:53 PM, Nicolas R. wrote:
> The problem of changing |Dumper| function behavior... is that it could 
> break maybe unexpected code trying to parse the output. The alternate 
> would be to provide an additional function (name to be defined) doing 
> the job.
> 

Yes, an additional function would probably be good.

**Comment by atoomic** [comment]
probably even better than a new function, a new $D:D::FLAG

**Comment by jkeenan** [comment]
@toddr I believe that you have reached some *modus vivendi* with Perl 5 Porters on this issue.  Could you just write up a summary of that for this ticket, so that we can then close it?

Thank you very much.
Jim Keenan

**Comment by toddr** [comment]
The changes to Dumper.pm are very minor. All we did was add my/or and called it a day. There is technically one change to behavior which is `my $pad`. This causes some of the generated code to be different and some of the tests reflect that. Other than this, the tests were just updated to accommodate strict.

As for the conflict, leon objected to the use of `add_my_to_dump` in any test and felt it should be removed. I removed the one nico added and replaced it with copious `no strict` injections. Then in a separate commit, I removed the existing 2 which people felt were a problem. ¯\_(ツ)_/¯

**Comment by jkeenan** [comment]
> The changes to Dumper.pm are very minor. All we did was add my/or and called it a day. There is technically one change to behavior which is `my $pad`. This causes some of the generated code to be different and some of the tests reflect that. Other than this, the tests were just updated to accommodate strict.
> 
> As for the conflict, leon objected to the use of `add_my_to_dump` in any test and felt it should be removed. I removed the one nico added and replaced it with copious `no strict` injections. Then in a separate commit, I removed the existing 2 which people felt were a problem. ¯_(ツ)_/¯

Merci beaucoup!  Closing ticket.
